File: app/src/main/java/bapspatil/captainchef/sync/UpdateRecipeService.java
package bapspatil.captainchef.sync;

import android.app.IntentService;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.Nullable;

import java.util.ArrayList;

import bapspatil.captainchef.model.Ingredient;

/**
 * Created by bapspatil
 */

public class UpdateRecipeService extends IntentService {

    public UpdateRecipeService() { super("UpdateRecipeService"); }

    // Helper method to explicitly start the UpdateRecipeService
    public static void startRecipeWidgetService(Context context, ArrayList<Ingredient> ingredientArrayList, String foodItemName) {
        Intent intent = new Intent(context, UpdateRecipeService.class);
        intent.putParcelableArrayListExtra("ingredientsList", ingredientArrayList);
        intent.putExtra("foodItemName", foodItemName);
        context.startService(intent);
    }

    @Override
    protected void onHandleIntent(@Nullable Intent intent) {
        if(intent != null) {
            String foodItemName = intent.getStringExtra("foodItemName");
            ArrayList<Ingredient> ingredientArrayList = intent.getParcelableArrayListExtra("ingredientsList");
            handleRecipeWidgetUpdate(foodItemName, ingredientArrayList);
        }
    }

    // Handle the recipe update action and send the broadcast to the WidgetProvider which updates the widget
    public void handleRecipeWidgetUpdate(String foodItemName, ArrayList<Ingredient> ingredientArrayList) {
        Intent intentToWidget = new Intent("android.appwidget.action.RECIPE_UPDATE");
        intentToWidget.setAction("android.appwidget.action.RECIPE_UPDATE");
        intentToWidget.putExtra("foodItemName", foodItemName);
        intentToWidget.putParcelableArrayListExtra("ingredientsList", ingredientArrayList);
        sendBroadcast(intentToWidget);
    }
}

File: app/src/main/java/bapspatil/captainchef/sync/UpdateRecipeService.kt
package bapspatil.captainchef.sync

import android.app.IntentService
import android.content.Context
import android.content.Intent
import bapspatil.captainchef.model.Ingredient
import java.util.*

/**
 * Created by bapspatil
 */

class UpdateRecipeService : IntentService("UpdateRecipeService") {

    override fun onHandleIntent(intent: Intent?) {
        if (intent != null) {
            val foodItemName = intent.getStringExtra("foodItemName")
            val ingredientArrayList = intent.getParcelableArrayListExtra<Ingredient>("ingredientsList")
            handleRecipeWidgetUpdate(foodItemName, ingredientArrayList)
        }
    }

    // Handle the recipe update action and send the broadcast to the WidgetProvider which updates the widget
    fun handleRecipeWidgetUpdate(foodItemName: String, ingredientArrayList: ArrayList<Ingredient>) {
        val intentToWidget = Intent("android.appwidget.action.RECIPE_UPDATE")
        intentToWidget.action = "android.appwidget.action.RECIPE_UPDATE"
        intentToWidget.putExtra("foodItemName", foodItemName)
        intentToWidget.putParcelableArrayListExtra("ingredientsList", ingredientArrayList)
        sendBroadcast(intentToWidget)
    }

    companion object {

        // Helper method to explicitly start the UpdateRecipeService
        fun startRecipeWidgetService(context: Context, ingredientArrayList: ArrayList<Ingredient>, foodItemName: String) {
            val intent = Intent(context, UpdateRecipeService::class.java)
            intent.putParcelableArrayListExtra("ingredientsList", ingredientArrayList)
            intent.putExtra("foodItemName", foodItemName)
            context.startService(intent)
        }
    }
}
